### PR TITLE
[plugin-json] Add patch json file step

### LINKED ISSUE #1476: new step: patch json
Sometimes it is necessary to update json variable by JSONPath: add, remove, replace field or field with value.

New step proposed:
```
When I save patched JSON `$varFrom` to `$scopes` variable `$varTo`:
|operation |path                |value                           |
|add       |$.path1             |["value"]                       |
|add       |$.path2             |"field": "value"                |
|remove    |$.path3             |                                |
|replace   |$.path4             |"field": "value"                |
|replace   |${pathSavedEarlier} |"field": "${valueSavedEarlier}" |
```

### PR DESCRIPTION
**Pull request by EDbarvinsky**
Closes #1476

### CODE REVIEW

File: vividus-plugin-json/build.gradle [hunk 1]
```diff
@@ -0,0 +1,18 @@
+project.description = 'Vividus plugin for JSON testing'
+
+dependencies {
+    implementation project(':vividus-bdd-engine')
+    implementation project(':vividus-soft-assert')
+    implementation project(':vividus-util')
+    implementation(group: 'com.jayway.jsonpath', name: 'json-path-assert', version: '2.4.0') {
+        exclude module: 'hamcrest-library'
+    }
+    implementation (group: 'net.javacrumbs.json-unit', name: 'json-unit', version: '2.18.0')
```
Comment (EDbarvinsky) [inline vividus-plugin-json/build.gradle:10]: Have tried to get rid of this dependencies, but do not succeed. Will discuss tomorrow

File: vividus-plugin-json/src/main/java/org/vividus/bdd/steps/json/JsonSteps.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+import java.nio.file.Paths;
+import java.util.Set;
+
+import com.fasterxml.jackson.databind.json.JsonMapper;
+import com.flipkart.zjsonpatch.JsonPatch;
+import org.jbehave.core.annotations.When;
+import org.vividus.bdd.context.IBddVariableContext;
+import org.vividus.bdd.variable.VariableScope;
+
+public class JsonSteps
```
Comment (valfirst) [inline vividus-plugin-json/src/main/java/org/vividus/bdd/steps/json/JsonSteps.java:16]: `JsonPatchSteps`

File: vividus-plugin-json/src/main/java/org/vividus/bdd/steps/json/JsonSteps.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+{
+    private final IBddVariableContext bddVariableContext;
+
+    public JsonSteps(IBddVariableContext bddVariableContext)
+    {
+        this.bddVariableContext = bddVariableContext;
+    }
+
+    @When("I patch content from json '$sourceJsonFilePath' with '$patchJsonFilePath' and put patched json to '$scopes'"
+            + " variable with name '$pathVariable'")
```
Comment (valfirst) [inline vividus-plugin-json/src/main/java/org/vividus/bdd/steps/json/JsonSteps.java:26]:
```
When I patch JSON `$sourceJson` using `$jsonPatch` and save result to $scopes variable `$variableName`
```

**Review by valfirst — CHANGES_REQUESTED** [review]

File: vividus-plugin-json/build.gradle [hunk 4]
```diff
@@ -0,0 +1,11 @@
+project.description = 'Vividus plugin for JSON testing'
+
+dependencies {
+    implementation project(':vividus-bdd-engine')
+    implementation (group: 'com.fasterxml.jackson.core', name: 'jackson-databind', version: '2.12.2')
```
Comment (valfirst) [inline vividus-plugin-json/build.gradle:5]: do you really need it?

File: vividus-plugin-json/src/main/java/org/vividus/bdd/steps/json/JsonPatchSteps.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    {
+        JsonNode jsonSource = readJson(sourceJson);
+        JsonNode patchJson = readJson(jsonPatch);
+        JsonNode patchedJson = JsonPatch.apply(patchJson, jsonSource);
+        bddVariableContext.putVariable(scopes, variableName, patchedJson.toString());
+    }
+
+    private JsonNode readJson(String json) throws JsonProcessingException
+    {
+        JsonMapper jsonMapper = new JsonMapper();
```
Comment (valfirst) [inline vividus-plugin-json/src/main/java/org/vividus/bdd/steps/json/JsonPatchSteps.java:51]: is it thread safe? if so, it should be a field of class

File: vividus-plugin-json/src/main/java/org/vividus/bdd/steps/json/JsonPatchSteps.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+public class JsonPatchSteps
+{
+    private final IBddVariableContext bddVariableContext;
+
+    public JsonPatchSteps(IBddVariableContext bddVariableContext)
+    {
+        this.bddVariableContext = bddVariableContext;
+    }
+
+    @When("I patch JSON `$sourceJson` using `$jsonPatch` and save result to $scopes variable `$variableName`")
```
Comment (valfirst) [inline vividus-plugin-json/src/main/java/org/vividus/bdd/steps/json/JsonPatchSteps.java:39]: please add javadocs

**Review by valfirst — CHANGES_REQUESTED** [review]
Please add:
- regular docs (see `/docs` folder)
- integration tests

File: settings.gradle [hunk 7]
```diff
@@ -33,6 +33,7 @@ include 'vividus-plugin-mobile-app'
 include 'vividus-plugin-parquet'
 include 'vividus-plugin-rest-api'
 include 'vividus-plugin-saucelabs'
+include 'vividus-plugin-json'
```
Comment (ikalinin1) [inline settings.gradle:36]: I believe we follow alphabetical order here

File: vividus-plugin-json/src/main/java/org/vividus/bdd/steps/json/JsonPatchSteps.java (hunk 6 above)
Comment (ikalinin1) [reply vividus-plugin-json/src/main/java/org/vividus/bdd/steps/json/JsonPatchSteps.java:39]: and documentation

File: docs/modules/plugins/pages/plugin-json.adoc [hunk 8]
```diff
@@ -0,0 +1,25 @@
+= Json Plugin
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-json.adoc:1]: let's use `JSON` notation everywhere

File: docs/modules/plugins/pages/plugin-json.adoc [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+== Steps
+
+=== Patch JSON
+
+[source,gherkin]
+----
+When I patch JSON `$sourceJson` using `$jsonPatch` and save result to $scopes variable `$variableName`
+----
+
+* `$sourceJson` - JSON file to be patched.
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-json.adoc:22]: it's not file, but JSON data

File: docs/modules/plugins/pages/plugin-json.adoc [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+* `$jsonPatch` - JSON file with required patch actions according to https://tools.ietf.org/html/rfc6902#page-6[RFC-6902]
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-json.adoc:23]: not file, similar to the above comment

File: docs/modules/plugins/pages/plugin-json.adoc [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+* `$scopes` - xref:parameters:variable-scope.adoc[the comma-separated set of the variables scopes]
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-json.adoc:24]: `The`

File: vividus-plugin-json/src/main/java/org/vividus/bdd/steps/json/JsonPatchSteps.java [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+
+    public JsonPatchSteps(IBddVariableContext bddVariableContext)
+    {
+        this.bddVariableContext = bddVariableContext;
+    }
+
+    /**
+     * Performs patching of a json and save it to the variable
+     * @param sourceJson json file to be patched
+     * @param jsonPatch json file with patch actions
```
Comment (valfirst) [inline vividus-plugin-json/src/main/java/org/vividus/bdd/steps/json/JsonPatchSteps.java:43]: please see related comments for the documentation

**Review by ikalinin1 — APPROVED** [review]

File: docs/modules/plugins/pages/plugin-json.adoc [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+* `$variableName` - The variable
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-json.adoc:25]: please add example of using step

Comment (valfirst) [inline docs/modules/plugins/pages/plugin-json.adoc:25]: `The variable name`

**Review by valfirst — CHANGES_REQUESTED** [review]

File: docs/modules/plugins/pages/plugin-json.adoc [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+----
+
+* `$sourceJson` - JSON data to be patched.
+* `$jsonPatch` - JSON data with required patch actions according to https://tools.ietf.org/html/rfc6902#page-6[RFC-6902]
+* `$scopes` - xref:parameters:variable-scope.adoc[The comma-separated set of the variables scopes]
+* `$variableName` - The variable name
+
+=== Examples
+
+.Send the data to the Kafka topic
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-json.adoc:29]: obviously it's about Kafka 😄

**Review by valfirst — CHANGES_REQUESTED** [review]

File: settings.gradle [hunk 15]
```diff
@@ -25,6 +25,7 @@ include 'vividus-plugin-db'
 include 'vividus-plugin-excel'
 include 'vividus-plugin-electron'
 include 'vividus-plugin-html'
+include 'vividus-plugin-json'
```
Comment (valfirst) [inline settings.gradle:28]: also please add the plugin here: https://github.com/vividus-framework/vividus/blob/master/vividus-docker-bundler/build.gradle

File: vividus-plugin-json/build.gradle [hunk 16]
```diff
@@ -0,0 +1,10 @@
+project.description = 'Vividus plugin for JSON testing'
+
+dependencies {
+    implementation project(':vividus-bdd-engine')
+    implementation (group: 'com.flipkart.zjsonpatch', name: 'zjsonpatch', version: '0.4.11')
+
+    testImplementation platform(group: 'org.junit', name: 'junit-bom', version: versions.junit)
+    testImplementation(group: 'org.junit.jupiter', name: 'junit-jupiter')
+    testImplementation(group: 'org.mockito', name: 'mockito-junit-jupiter', version: versions.mockito)
+}
```
Comment (valfirst) [inline vividus-plugin-json/build.gradle:10]: also please add these 2 lines here at the end of file:
https://github.com/vividus-framework/vividus/blob/057c1f41cff1b2038b1979bde91b126f0da6e3f6/vividus-plugin-websocket/build.gradle#L16-L17

**Review by valfirst — APPROVED** [review]

**Review by uarlouski — APPROVED** [review]

**Pull request merged**